pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/automaton/RunAutomaton.java

```diff
@@ -0,0 +1,311 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils.nativefst.automaton;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.net.URL;
+import java.util.Set;
+
+
+/**
+ * Finite-state automaton with fast run operation.
+ */
+public class RunAutomaton implements Serializable {
+
+  int _size;
+  boolean[] _accept;
+  int _initial;
+  int[] _transitions; // delta(state,c) = transitions[state*points.length + getCharClass(c)]
+  char[] _points; // char interval start points
+  int[] _classmap; // map from char number to class class
+
+  /**
+   * Constructs a new <code>RunAutomaton</code> from a deterministic
+   * <code>Automaton</code>. Same as <code>RunAutomaton(a, true)</code>.
+   * @param a an automaton
+   */
+  public RunAutomaton(Automaton a) {
+    this(a, true);
+  }
+
+  /**
+   * Constructs a new <code>RunAutomaton</code> from a deterministic
+   * <code>Automaton</code>. If the given automaton is not deterministic,
+   * it is determinized first.
+   * @param a an automaton
+   * @param tableize if true, a transition table is created which makes the <code>run</code>
+   *                 method faster in return of a higher memory usage
+   */
+  public RunAutomaton(Automaton a, boolean tableize) {
+    a.determinize();
+    _points = a.getStartPoints();
+    Set<State> states = a.getStates();
+    Automaton.setStateNumbers(states);
+    _initial = a._initial._number;
+    _size = states.size();
+    _accept = new boolean[_size];
+    _transitions = new int[_size * _points.length];
+    for (int n = 0; n < _size * _points.length; n++) {
+      _transitions[n] = -1;
+    }
+    for (State s : states) {
+      int n = s._number;
+      _accept[n] = s._accept;
+      for (int c = 0; c < _points.length; c++) {
+        State q = s.step(_points[c]);
+        if (q != null) {
+          _transitions[n * _points.length + c] = q._number;
+        }
+      }
+    }
+    if (tableize) {
+      setAlphabet();
+    }
+  }
+
+  /**
+   * Retrieves a serialized <code>RunAutomaton</code> located by a URL.
+   * @param url URL of serialized automaton
+   * @exception IOException if input/output related exception occurs
+   * @exception ClassCastException if the data is not a serialized <code>RunAutomaton</code>
+   * @exception ClassNotFoundException if the class of the serialized object cannot be found
+   */
+  public static RunAutomaton load(URL url)
+      throws IOException, ClassCastException, ClassNotFoundException {
+    return load(url.openStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *openStream*. Currently, there are execution paths that do not contain closure statements, for example, when RunAutomaton.load throws an exception. Ensure the object returned by *openStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).